File: Day Four 23-10-24/solution-1.ts
const isValidParentheses = (s: string): boolean => {
  // Initialize an empty stack to keep track of opening brackets
  const stack: string[] = [];
  // Map to store the matching pairs of brackets
  const bracketPairs: { [key: string]: string } = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  // Iterate over each character in the string
  for (let i = 0; i < s.length; i++) {
    const char: string = s[i]; // Current character
    // If the character is an opening bracket, push it onto the stack
    if (map[char]) {
      stack.push(char); // Use lowercase 'stack'
    } else {
      // If the character is a closing bracket, pop the top element from the stack
      const topElement: string | undefined = stack.pop(); // Use lowercase 'stack'
      // Check if the popped element matches the current closing bracket
      if (!topElement || map[topElement] !== char) {
        // Added check for undefined topElement
        return false; // If not, the string is not valid
      }
    }
  }

  // If the stack is empty, all brackets were matched correctly
  return stack.length === 0; // Use lowercase 'stack'
};

File: Day Four 23-10-24/solution-1.py
class Solution:
    def isValid(self, s):
        # Initialize an empty list to use as a stack for tracking opening brackets
        stack = []
        # Dictionary to map each type of opening bracket to its corresponding closing bracket
        map = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        
        # Iterate over each character in the input string
        for char in s:
            # If the character is an opening bracket, push it onto the stack
            if char in map:
                stack.append(char)
            else:
                # If the stack is empty, it means there's no matching opening bracket
                if not stack:
                    return False
                # Pop the top element from the stack, which should be the last unmatched opening bracket
                topElement = stack.pop()
                # Check if the popped element's corresponding closing bracket matches the current character
                if map[topElement] != char:
                    return False
        
        # If the stack is empty, all brackets were matched correctly; otherwise, return False
        return not stack

# Test the function with a sample input
solution = Solution()
print(solution.isValid("()"))